### DISCUSSION TITLE
How to mount to lowerdir (overlayfs)

### DISCUSSION THREAD

**Discussion by dmpopoff** [moby/moby#45175] [Q&A] [answered]
I need to include some PHP code in the docker image and at the same time use some code from host directory while maintaining a certain file structure inside the running container.

How to write mount of hostdir1 in docker-compose.yml file correctly?

Let me explain with a short example:

1. inside image (after build):

```
tree /mount
mount/
├── 1.txt
├── 4.txt
└── 5.txt
 cat mount/1.txt 
 internal
```

2. on the host filesystem:

```
hostdir1/
├── 1.txt
├── 2.txt
└── somedir
       └── 3.txt
 cat hostdir1/1.txt
 external
```

3. I need to mount host folder to contaider using smth like overlayFS and get the following file structure in the running container:

```
mount/
├── 1.txt
├── 2.txt
├── 4.txt
├── 5.txt
└── somedir
    └── 3.txt
cat mount/1.txt 
internal
```
[Here](https://stackoverflow.com/questions/40905761/how-do-i-mount-a-host-directory-as-a-volume-in-docker-compose) they simply mount the directory into the container, overwriting all the contents of the directory in the image. I need to mount the host directory from below to the desired image directory. Keeping available all the files that are in the image and adding to them those files and directories that are on the server but are not in the image.

**Comment by dmpopoff** [comment]
@amir73il please look at the situation. Could you help with an answer?

**Reply by amir73il** [reply]
I am assuming you want the image to be the upperdir so you can write to the image?
so your problem is where to put the workdir?
Is it not possible for you to create the image files under a "root" dir?
or mount the image and move all the files/dirs under / into /root/?
then you can create a /work dir sibling to /root and mount the overlay.
or maybe I don't understand the problem?

**Reply by dmpopoff** [reply]
I can create mountpoints directories as "upperdir" "work" and "merged" inside the image in / (root). It`s not a problem. I dont understand syntax of mount construstion (when I run container) if I want mount hostdir as a lowerdir.

for example I can use simple 
docker run -v volume1:/dir1 myimage:latest
but in this case content from "volume1" is mounted as upperdir. I need to mount it as a lowerdir and image dir "/dir1" as an upperdir.

**Reply by amir73il** [reply]
It is a standard overlayfs scenario

mount -t overlay -olowerdir=/hostdir1,upperdir=/image/root,workdir=/image/work /merge

what am I missing?

**Reply by dmpopoff** [reply]
Yes, it is standard scenario. Could you help - how I can use your construction in **docker run** command?

**Reply by dmpopoff** [reply]
confuses that in my understanding mounting has to be done at the moment of creation of the container. And if we look at the mount [documentation](https://docs.docker.com/storage/bind-mounts/), there are no options that you specify. There is "type=", there is "target=", but there is no "-t overlay -olowerdir=", for example.

**Reply by amir73il** [reply]
I dont know docker. you need to ask the docker developers.

**Reply by dmpopoff** [reply]
thank you for your participation

**Comment by dmpopoff** [comment, marked as answer]
@corhere could you please suggest a solution? need to mount the hostdir as a lowerdir to the starting container.

**Reply by dmpopoff** [reply]
files in the upperdir (in the image) don't care about saving changes.

**Reply by corhere** [reply]
I suggest [you look for another way to solve your problem](https://xyproblem.info/) as there is no supported way to start a container with a bind-mount where the host directory is an overlayfs lowerdir.

**Reply by dmpopoff** [reply]
Thanks a lot. To understand that this path is a dead end means to stop wasting time here and make efforts in other options. I`ll try solve my problem another way:)
